injection/sharedmain/main.go

```diff
@@ -93,74 +93,128 @@ func GetLoggingConfig(ctx context.Context) (*logging.Config, error) {
 	return logging.NewConfigFromConfigMap(loggingConfigMap)
 }
 
+// Main runs the generic main flow for non-webhook controllers with a new
+// context. Use WebhookMainWith* if you need to serve webhooks.
 func Main(component string, ctors ...injection.ControllerConstructor) {
 	// Set up signals so we handle the first shutdown signal gracefully.
 	MainWithContext(signals.NewContext(), component, ctors...)
 }
 
+// MainWithContext runs the generic main flow for non-webhook controllers. Use
+// WebhookMainWithContext if you need to serve webhooks.
 func MainWithContext(ctx context.Context, component string, ctors ...injection.ControllerConstructor) {
-	var (
-		masterURL = flag.String("master", "",
-			"The address of the Kubernetes API server. Overrides any value in kubeconfig. Only required if out-of-cluster.")
-		kubeconfig = flag.String("kubeconfig", "",
-			"Path to a kubeconfig. Only required if out-of-cluster.")
-	)
-	flag.Parse()
-
-	cfg, err := GetConfig(*masterURL, *kubeconfig)
-	if err != nil {
-		log.Fatalf("Error building kubeconfig: %v", err)
-	}
+	cfg := ParseAndGetConfigOrDie()
 	MainWithConfig(ctx, component, cfg, ctors...)
 }
 
+// MainWithConfig runs the generic main flow for non-webhook controllers. Use
+// WebhookMainWithConfig if you need to serve webhooks.
 func MainWithConfig(ctx context.Context, component string, cfg *rest.Config, ctors ...injection.ControllerConstructor) {
 	log.Printf("Registering %d clients", len(injection.Default.GetClients()))
 	log.Printf("Registering %d informer factories", len(injection.Default.GetInformerFactories()))
 	log.Printf("Registering %d informers", len(injection.Default.GetInformers()))
 	log.Printf("Registering %d controllers", len(ctors))
 
-	// Report stats on Go memory usage every 30 seconds.
-	msp := metrics.NewMemStatsAll()
-	msp.Start(ctx, 30*time.Second)
+	MemStatsOrDie(ctx)
 
-	if err := view.Register(msp.DefaultViews()...); err != nil {
-		log.Fatalf("Error exporting go memstats view: %v", err)
+	// Adjust our client's rate limits based on the number of controller's we are running.
+	cfg.QPS = float32(len(ctors)) * rest.DefaultQPS
+	cfg.Burst = len(ctors) * rest.DefaultBurst
+	ctx, informers := injection.Default.SetupInformers(ctx, cfg)
+
+	// Set up our logger.
+	logger, atomicLevel := SetupLoggerOrDie(ctx, component)
+	defer flush(logger)
+
+	// Obtain K8s clientset.
+	kc := kubeclient.Get(ctx)
+	if err := version.CheckMinimumVersion(kc.Discovery()); err != nil {
+		logger.Fatalw("Version check failed", zap.Error(err))
+	}
+
+	// Create configMap watcher
+	cmw := SetupConfigMapWatchOrDie(ctx, logger)
+
+	// Based on the reconcilers we have linked, build up the set of controllers to run.
+	controllers := make([]*controller.Impl, 0, len(ctors))
+	for _, cf := range ctors {
+		ctrl := cf(ctx, cmw)
+		controllers = append(controllers, ctrl)
 	}
 
+	profilingHandler := profiling.NewHandler(logger, false)
+
+	// Watch the logging config map and dynamically update logging levels.
```

Comment:
maybe move this  and observability die above setting up the controllers so it's grouped by where the pieces necessary to construct it are set up? My reasoning being that if all the dies are as early as possible, just makes it easier to reason about the code?
Again, not a big deal :)

Reply:
Ack, I tried to preserve the existing flow as much as possible for lower review overhead, but reading it now it's a bit apart. I'll rejigger